core/state/access_events.go

```diff
@@ -0,0 +1,314 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package state
+
+import (
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/math"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/ethereum/go-ethereum/trie/utils"
+	"github.com/holiman/uint256"
+)
+
+// mode specifies how a tree location has been accessed
+// for the byte value:
+// * the first bit is set if the branch has been edited
+// * the second bit is set if the branch has been read
+type mode byte
+
+const (
+	AccessWitnessReadFlag  = mode(1)
+	AccessWitnessWriteFlag = mode(2)
+)
+
+var zeroTreeIndex uint256.Int
+
+// AccessEvents lists the locations of the state that are being accessed
+// during the production of a block.
+type AccessEvents struct {
+	branches map[branchAccessKey]mode
+	chunks   map[chunkAccessKey]mode
+
+	pointCache *utils.PointCache
+}
+
+func NewAccessEvents(pointCache *utils.PointCache) *AccessEvents {
+	return &AccessEvents{
+		branches:   make(map[branchAccessKey]mode),
+		chunks:     make(map[chunkAccessKey]mode),
+		pointCache: pointCache,
+	}
+}
+
+// Merge is used to merge the access events that were generated during the
+// execution of a tx, with the accumulation of all access events that were
+// generated during the execution of all txs preceding this one in a block.
+func (ae *AccessEvents) Merge(other *AccessEvents) {
+	for k := range other.branches {
+		ae.branches[k] |= other.branches[k]
+	}
+	for k, chunk := range other.chunks {
+		ae.chunks[k] |= chunk
+	}
+}
+
+// Keys returns, predictably, the list of keys that were touched during the
+// buildup of the access witness.
+func (ae *AccessEvents) Keys() [][]byte {
+	// TODO: consider if parallelizing this is worth it, probably depending on len(ae.chunks).
+	keys := make([][]byte, 0, len(ae.chunks))
+	for chunk := range ae.chunks {
+		basePoint := ae.pointCache.Get(chunk.addr[:])
+		key := utils.GetTreeKeyWithEvaluatedAddress(basePoint, &chunk.treeIndex, chunk.leafKey)
+		keys = append(keys, key)
+	}
+	return keys
+}
+
+func (ae *AccessEvents) Copy() *AccessEvents {
+	cpy := &AccessEvents{
+		branches:   make(map[branchAccessKey]mode),
+		chunks:     make(map[chunkAccessKey]mode),
+		pointCache: ae.pointCache,
+	}
+	cpy.Merge(ae)
+	return cpy
+}
+
+// AddAccount returns the gas to be charged for each of the currently cold
+// member fields of an account.
+func (ae *AccessEvents) AddAccount(addr []byte, isWrite bool) uint64 {
+	var gas uint64
+	for i := utils.VersionLeafKey; i <= utils.CodeSizeLeafKey; i++ {
+		gas += ae.touchAddressAndChargeGas(addr, zeroTreeIndex, byte(i), isWrite)
+	}
```

Comment:
I think you should spell it out explicitly. The loop here doesn't improve readability
```suggestion
	gas += ae.touchAddressAndChargeGas(addr, zeroTreeIndex, VersionLeafKey, isWrite)
	gas += ae.touchAddressAndChargeGas(addr, zeroTreeIndex, BalanceLeafKey, isWrite)
	gas += ae.touchAddressAndChargeGas(addr, zeroTreeIndex, NonceLeafKey, isWrite)
	gas += ae.touchAddressAndChargeGas(addr, zeroTreeIndex, CodeKeccakLeafKey, isWrite)
	gas += ae.touchAddressAndChargeGas(addr, zeroTreeIndex, CodeSizeLeafKey, isWrite)
```
(same goes for all places where it looks like this)